### PR TITLE
add command genkey

### PR DESCRIPTION
**Pull request by behrmann**
First draft to address #623.

This leaves out any checking whether the keys are in git or whether they or have wrong permissions, since openssl created them with the right ones.

I'm not entirely sure what a good choice for the CN or the expiration length of the keys should be (ten years might be a tad long), hence this is a draft. I haven't yet generated my own secure boot keys, so I'm also a bit hazy the exact incantations, i.e. if longer keys than 2048 bit are allowed (I remember reading somewhere that they are not) and that to put them into EFI variables, the cert needs to be converted from PEM into DER, so maybe we should do this as well.

Happy about comments.

### CODE REVIEW

File: mkosi/__init__.py [hunk 1]
```diff
@@ -6273,6 +6273,37 @@ def run_ssh(args: CommandLineArguments) -> None:
...
+            "-new",
+            "-x509",
+            "-newkey",
+            "rsa:2048",
+            "-keyout",
+            args.secure_boot_key or "./mkosi.secure-boot.key",
+            "-out",
+            args.secure_boot_certificate or "./mkosi.secure-boot.crt",
+            "-days",
+            "3650",
```
Comment (poettering) [inline mkosi/__init__.py:6300]: i'd probably do 2y. i.e. 1y sounds like a good time to roll them over, and let's add a bit of space around that, and doubling sounds like a safe choice for that.

File: mkosi/__init__.py [hunk 2]
```diff
@@ -6273,6 +6273,37 @@ def run_ssh(args: CommandLineArguments) -> None:
...
+            "-subj",
+            "/CN=mkosi/",
```
Comment (poettering) [inline mkosi/__init__.py:6302]: I'd probably include the local user name. i.e. maybe generate `mkosi of lennart` for a user lennart  or something like that.

I think the validity and the CN are something that might be worth making configurable in new `mkosi.default` keys.

**Review by poettering — CHANGES_REQUESTED** [review]
looks pretty good to me. I think 2K RSA sounds like a good default. Until someone asks for something else I don't think we need to make that configurable. i.e. i'd hardcode this for now until someone complains.

PEM should be the right choice for encoding the key material. If some tools need DER, then we this should be converted on-the-fly, but never be stored in DER persistently ever.

**Comment by poettering** [comment]
maybe while generating them tell user they need to rollover in the configured time, and the CN that was chosen.

File: mkosi/__init__.py (hunk 1 above)
Comment (behrmann) [reply mkosi/__init__.py]: Changed the time to two years and added a note to the user, when the key expires.

File: mkosi/__init__.py (hunk 2 above)
Comment (behrmann) [reply mkosi/__init__.py]: Added a note about the CN, which defaults to `mkosi of user`, where user goes through `SUDO_USER`, `USER` and defaults to `unknown user`.

I've so far decided not to add any configurability.

**Comment by behrmann** [comment]
@DaanDeMeyer Currently our CI always seems to fail for arch `gpt_ext4` and `gpt_xfs` due to `systemd-tmpfiles-setup.service` hanging. Any ideas?

**Comment by DaanDeMeyer** [comment]
https://github.com/systemd/systemd/issues/18140

File: mkosi/__init__.py [hunk 3]
```diff
@@ -6285,6 +6286,51 @@ def run_ssh(args: CommandLineArguments) -> None:
...
+    user = os.getenv("SUDO_USER") or os.getenv("USER") or "unknown user"
+    cn = f"mkosi of {user}"
+
+    for f in (args.secure_boot_key, args.secure_boot_certificate):
+        if os.path.exists(f):
+            die(
+                f"{f} already exists.\n"
+                "If you are sure you want to generate new secure boot keys\n"
+                f"remove {args.secure_boot_key} and {args.secure_boot_certificate} first."
+            )
```
Comment (poettering) [inline mkosi/__init__.py:6306]: this check should probably be skipped if --force is specified (in which case the key files should just be replaced)

**Comment by poettering** [comment]
So, would still love to see the cn and the days configureable in the mkosi file, and the --force hookup. Otherwise looks excellent to me

**Comment by behrmann** [comment]
This version makes the validity of the keys and the CN configurable. It also adds a very rudimentary `expand_specifier` function, that's not yet hooked up to the config parsing mechanism, so that CN can be `mkosi of %u` and expand to the user. This comes with the downside of having to disable the interpolation that comes with `configparser.ConfigParser` by switching to `configparser.RawConfigParser` instead. This is a backwards-incompatible change and might break somebody who used this configparser feature. It was never advertised, but there's the off chance that somebody did. I hope nobody did, at least I always forget, that it exists.

I also added an unrelated commit on top to satisfy mypy, since the new version released yesterday fixed a long-standing bug that made a cast unnecessary.

**Comment by behrmann** [comment]
I'm not entirely happy with the new secure boot related options in the `[Output]` section, but I guess it's better to keep them close to the other secure boot related options, than start a section solely for them.

File: mkosi/__init__.py [hunk 4]
```diff
@@ -6285,6 +6298,49 @@ def run_ssh(args: CommandLineArguments) -> None:
...
+    keylength = 2048
+    expiration_date = datetime.date.today() + datetime.timedelta(int(args.secure_boot_valid_days))
+    cn = expand_specifier(args.secure_boot_common_name)
+
+    for f in (args.secure_boot_key, args.secure_boot_certificate):
+        if os.path.exists(f) and not args.force:
+            die(
+                f"{f} already exists.\n"
+                "If you are sure you want to generate new secure boot keys\n"
+                f"remove {args.secure_boot_key} and {args.secure_boot_certificate} first."
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:6364]: I'd make this a multiline string and use dedent

File: mkosi/__init__.py [hunk 5]
```diff
@@ -6285,6 +6298,49 @@ def run_ssh(args: CommandLineArguments) -> None:
...
+            )
+
+    MkosiPrinter.print_step(f"Generating secure boot keys rsa:{keylength} for CN `{cn}`.")
+    MkosiPrinter.info(
+        f"The keys will expire in {args.secure_boot_valid_days} days ({expiration_date:%A %d. %B %Y}). "
+        "Remember to roll them over to new ones before."
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:6370]: before => before then?

File: mkosi/__init__.py [hunk 6]
```diff
@@ -4214,7 +4215,7 @@ def ini_key_to_cli_arg(key: str) -> str:
                 continue
             # replace arguments referencing files with the file content
             try:
-                config = configparser.ConfigParser(delimiters="=")
+                config = configparser.RawConfigParser(delimiters="=")
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:4262]: What's the motivation behind this change?

File: mkosi/__init__.py [hunk 7]
```diff
@@ -6285,6 +6298,49 @@ def run_ssh(args: CommandLineArguments) -> None:
         )
 
 
+def generate_secure_boot_key(args: CommandLineArguments) -> NoReturn:
+    """Generate secure boot keys using openssl"""
+    args.secure_boot_key = args.secure_boot_key or "./mkosi.secure-boot.key"
+    args.secure_boot_certificate = args.secure_boot_certificate or "./mkosi.secure-boot.crt"
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:6353]: Do these end up in the output directory if it is defined?

File: mkosi/__init__.py (hunk 4 above)
Comment (behrmann) [reply mkosi/__init__.py:6364]: Will do, but as long as it's only a few lines, I tend to prefer implicit concatenation instead of the additional level of indentation.

File: mkosi/__init__.py (hunk 6 above)
Comment (behrmann) [reply mkosi/__init__.py:4262]: The problem is that `ConfigParser` comes with its own interpolation facilities, that unfortunately clash with the format of systemd specifiers, because they too use %, e.g. `%(key)` expands to the value that key has if it was defined before in the config file. I also made a note of this in the commit message changing this.

Since @poettering noted in #626 that it would need some basic variable expansion, I went for `%u` to expand to user, so that specifier knowledge carries over from systemd. 

Unfortunately this change is backwards-incompatible, but I also don't think this is a terribly good feature to have for the mkosi config files as it was. That being said, I hope nobody uses this at the moment.

File: mkosi/__init__.py (hunk 7 above)
Comment (behrmann) [reply mkosi/__init__.py:6353]: Good question. I forgot to test that. The question is: should they? I think they should end up such, that mkosi can pick them up right away without setting `--secure-boot-{key,certificate}`.

File: mkosi/__init__.py (hunk 6 above)
Comment (DaanDeMeyer) [reply mkosi/__init__.py:4262]: I doubt it, LGTM

File: mkosi/__init__.py (hunk 7 above)
Comment (DaanDeMeyer) [reply mkosi/__init__.py:6353]: I think they should, we do it for all the other stuff so I don't see a big reason to deviate from that here.

Comment (behrmann) [reply mkosi/__init__.py:6356]: Ok, I'll have a look at it tomorrow.

File: mkosi/__init__.py [hunk 8]
```diff
@@ -4258,7 +4259,10 @@ def ini_key_to_cli_arg(key: str) -> str:
                 continue
             # replace arguments referencing files with the file content
             try:
-                config = configparser.ConfigParser(delimiters="=")
+                # This used to us configparser.ConfigParser before, but
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:4262]: us => use?

File: mkosi/__init__.py [hunk 9]
```diff
@@ -6334,6 +6350,57 @@ def run_ssh(args: CommandLineArguments) -> None:
...
+        )
+    )
+
+    cmd = [
+        "openssl",
+        "req",
+        "-new",
+        "-x509",
+        "-newkey",
+        f"rsa:2048",
```
Comment (DaanDeMeyer) [inline mkosi/__init__.py:6390]: Need to use `keylength` here?

**Review by DaanDeMeyer — COMMENTED** [review]
One more question, how does this integrate with the existing secureboot stuff? Will the keys generated here automatically be used to sign the image if building with mkosi afterwards?

File: mkosi/__init__.py (hunk 9 above)
Comment (behrmann) [reply mkosi/__init__.py]: Thanks! That's a holdover from an earlier revision.

File: mkosi/__init__.py (hunk 7 above)
Comment (behrmann) [reply mkosi/__init__.py:6356]: So currently this would ignore `args.output_dir`, but to answer the question below, there they wouldn't be picked up automatically, since `find_secure_boot` looks relative to `args.directory`.

I'm still not sure whether supporting whether moving this relative to the output directory makes sense. `--secure-boot-key` and `--secure-boot-certificate` take paths that are not necessarily relative to the output directory. If somebody would set `--output-dir` and a relative path for either of the secure boot options the result could be confusing. 

I'm not sure what the best option is, but I think it would be great if they would be just be picked up for `--secure-boot`, which should usually be the case, if people just run the command at the root of their config. What do you think?

Comment (DaanDeMeyer) [reply mkosi/__init__.py:6356]: We can do it this way, thinking further, these keys are not something you'll be regenerating again and again anyway so it kinda makes sense that we just output them in the working directory instead of the output dir.
